client/network/src/service.rs

```diff
@@ -560,6 +568,71 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		});
 	}
 
+	/// Appends one or more notifications to the buffer of pending outgoing notifications with the
+	/// given peer.
+	///
+	/// The returned `Future` finishes only after all the notifications have been queued or the
+	/// substream has been closed.
+	///
+	/// This operation is atomic in the sense that either all or none of the notifications have
+	/// been enqueued. In particular, it is guaranteed that no notifications have been transmitted
+	/// if the operation is interrupted by dropping the `Future`.
+	///
+	/// An error is returned if there exists no open notifications substream with that combination
+	/// of peer and protocol, or if the remote has asked to close the notifications substream.
+	///
+	/// If the remote requests to close the notifications substream, all notifications successfully
+	/// enqueued with this method will finish being sent out before the substream actually gets
+	/// closed, but attempting to enqueue more notifications will now return an error. It is also
+	/// possible for the entire connection to be abruptly closed, in which case enqueued
+	/// notifications will be lost.
+	///
+	/// The protocol must have been registered with `register_notifications_protocol` or
+	/// `NetworkConfiguration::notifications_protocols`.
+	///
+	/// # Usage
+	///
+	/// This method waits until there space is available in the buffer of messages towards the
+	/// given peer. If the peer processes notifications at a slower rate than we send them, this
+	/// buffer will quickly fill up.
+	///
+	/// As such, you should never do something like this:
+	///
+	/// ```ignore
+	/// // Do NOT do this
+	/// for peer in peers {
+	///     network.send_notifications(peer, ..., notifications).await;
```

Comment:
I get that I'm not supposed to do this, but it's also kind of hard to manage potentially complex sending logic across many subsystems. 

@infinity0 suggested that it might be best for each subsystem to get a different substream for each peer to manage that.

In Polkadot we have this "Network Bridge" which I can probably refactor away, but it was nice for serving as an additional layer of abstraction between the "raw" network service and the protocols we are using.

I mentioned in [this comment](https://github.com/paritytech/substrate/issues/5481#issuecomment-654417083) that I'd like to use something like FuturesUnordered for this, so instead we'd be waiting on `network.send_notifications` for _all_ peers at once. That is OK?

Reply:
I'm not 100% sure what the expected control-flow is in this all-or-nothing style.

So if a peer has no space in their buffer for a message, I can take two immediate actions:
  * Remove a message from the buffer
  * Disconnect the peer

I might also like to defer until the peer has some space in their buffer and then attempt to send. While this is happening I cannot process incoming messages because the outcome of handling some incoming message might lead to us attempting yet another send to the peer, which I'm not sure how to handle. Because it's not clear how to handle multiple sends where there is no buffer, I prefer just to make this an impossibility.

Waiting for the peer to have space in their buffer should obviously have a timeout after which we just disconnect the peer because they are slowing everything down. But it seems like things actually slow down more with the all-or-nothing approach - if we only need `n` spaces in the buffer, we shouldn't have to wait for the whole buffer to be sent before buffering the next message.

Reply:
This function is not meant to be used by high-level code, it is the foundation on top of which we can implement something like `sc-network-gossip`, or Ximin's priority queue system, or something similar.

The flow control is meant to be: you push your messages to a queue (which isn't handled by `sc-network`), and this queue is, in parallel, processed by another task that sends out the messages to the various peers.
If a peer is too slow, then it is the role of this other task to stop sending to it and disconnect it.

> Zooming out a bit, it's also unclear how you would write the analogue of this code:

You'd do `some_channel.send(msg()).await;`, and the receiving side of the channel then attempts, in parallel, to send the message to all peers.

> I'm not 100% sure what the expected control-flow is in this all-or-nothing style.

The way you are supposed to use this API is really one message at a time.
The all-or-nothing system when passing multiple messages is something I'm not sure is actually useful.

Reply:
> The way you are supposed to use this API is really one message at a time.
The all-or-nothing system when passing multiple messages is something I'm not sure is actually useful.

If you mean one message at a time per-peer, then yes, I agree with you.

But if you mean one message at a time overall, then I am not sure if that is realistic.

However note that my example above is only about sending one message at a time per-peer. The all-or-nothing is about how the buffer drains, not anything else.

Reply:
> You'd do some_channel.send(msg()).await;, and the receiving side of the channel then attempts, in parallel, to send the message to all peers.

OK, this is fine, and actually suits the code we wrote for Polkadot networking very well already. However as soon as you introduce this layer of abstraction where the thing that actually calls `send_notifications` is one step removed from the thing which generates the messages, you lose the ability to meaningfully inspect and prune the buffer without callbacks.

Reply:
> However as soon as you introduce this layer of abstraction where the thing that actually calls send_notifications is one step removed from the thing which generates the messages, you lose the ability to meaningfully inspect and prune the buffer without callbacks.

This remains to be designed, but I believe it should be possible to use a `futures::lock::Mutex` to access the content of the queue.

One thing that we might have to change is that `send_notifications` should return an intermediate object, so that we don't have to immediately pass the notification by value. I'm going to adjust the PR.

Reply:
This is a very rough draft, but I have something like this in mind: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=427a0f40cd2d3e9bf9c3f5a228111045

Pushing a message to the queue would lock the queue. The background task would also lock that queue but only for a very brief moment in order to pop a message.

Reply:
OK, I will say that it needs a lot of work before it can be used in our gossip protocols, which are pseudo-gossip and rely on data-dependency - it's not a simple case of "every peer gets every message".